### ISSUE TITLE
Proxy CONNECT uses IP address rather than domain name

### ISSUE THREAD

**Issue by rjaffey** [jay0lee/GAM#45]
My organization uses a proxy for outbound HTTP and HTTPS access to internet resources from a server farm.  Access through the proxy is only authorized to specific domain names (e.g. .google.com, .googleapis.com, .apple.com, .microsoft.com, etc.).  This works fine for clients that issue

"CONNECT &lt;host.domain.name&gt;:&lt;port&gt;"

requests in order to tunnel HTTPS connections through the proxy.  With exception of GAM, all of our clients issue "CONNECT" this way.  In GAM's case

   httplib2/socks.py

builds

"CONNECT &lt;ipaddress&gt;:&lt;port&gt;"

requests instead.  This makes it challenging to limit destinations using our proxy server implementation, as we would have to allow access to all of Google's address space in order for GAM to work.  It would therefore be desirable for GAM to CONNECT via domain name if possible.

**Comment by rjaffey** [comment]
I'm sure this is not a proper fix, but the following patch addresses

**Comment by rjaffey** [comment]
Sorry, hit the wrong button, did not mean to close.

The following patch addresses my immediate need, but it's likely not a proper fix for general use:

--- GAM-3.42/httplib2/socks.py	2014-11-19 14:02:09.000000000 -0600
+++ ../GAM-3.42/httplib2/socks.py	2015-03-16 15:01:11.823531079 -0500
@@ -364,7 +364,7 @@
             addr = socket.gethostbyname(destaddr)
         else:
             addr = destaddr
-        headers =  ["CONNECT ", addr, ":", str(destport), " HTTP/1.1\r\n"]
+        headers =  ["CONNECT ", destaddr, ":", str(destport), " HTTP/1.1\r\n"]
         headers += ["Host: ", destaddr, "\r\n"]
         if (self.__proxy[4] != None and self.__proxy[5] != None):
                 headers += [self.__getauthheader(), "\r\n"]

**Comment by jay0lee** [comment]
httplib2 is upstream of GAM and would need to be patched there:

**Comment by jay0lee** [comment]
https://github.com/jcgregorio/httplib2

**Comment by rjaffey** [comment]
Actually, it looks like the version of httplib2 included in GAM and how GAM calls it that affects this:

The following code is present in the version of httplib2/__init__.py included in GAM 3.42:

```
    def __init__(self, proxy_type, proxy_host, proxy_port,
                 proxy_rdns=None, proxy_user=None, proxy_pass=None):
```
Whereas the following code is present in the current master branch of httplib2/__init__.py:

```
    def __init__(self, proxy_type, proxy_host, proxy_port,
                 proxy_rdns=True, proxy_user=None, proxy_pass=None):
```
along with the comment:
```
          proxy_rdns: If True (default), DNS queries will not be performed
          locally, and instead, handed to the proxy to resolve.  This is useful
          if the network does not allow resolution of non-local names.  In
          httplib2 0.9 and earlier, this defaulted to False.
```
I'd say, either updating the version of httplib2 or overriding the default when ProxyInfo is initialized would address the issue.  In either case this would point back at GAM as the place to fix this.
